### ISSUE TITLE
Print JS Print Items with Rust - "Write Item" IR

### ISSUE THREAD

**Issue by dsherret** [dsherret/dprint#52]
I have a working prototype of a printer written in Rust here: https://github.com/dsherret/dprint/tree/master/packages/rust-core

It would be nice to experiment printing JS print items with the Rust printer to see if a performance improvement could be gained (using Wasm). I looked into this and it seems the bottle neck will be sending over strings over to Rust (since strings in Rust are UTF-8 and strings in JS are UTF-16 they need to be encoded then decoded).

The amazing thing is that the Rust printer could be redesigned so the strings remain in JS **since the printer only really cares about string lengths**. It seems using [js_sys::String](https://rustwasm.github.io/docs/wasm-bindgen/reference/types/str.html#utf-16-vs-utf-8) allows that.

Basically, instead of strings, the writer (used internally by the printer) would build up another IR called "write items":

* `String` - Represented by an object containing a reference to the JS string and holds the string length. The string value MUST not contain tabs or newlines... if it does then the provided string should be broken up before its reference is sent to Rust.
* `Indent` (`u8` value) - A single indent.
* `NewLine` (`u8` value) - A newline character.
* `Tab` (`u8` value) - A tab character. This is different than an indent as an indent may not be a tab. Tabs need their own item in the IR because they increase the column number by a value greater than 1 (ex. a tab increases the column number by 4 when the indent width is set to 4).

The writer will then return this final write item IR to JS. When JS has it, it will then use a simple write item printer to print out the final string.

To compare the performance fairly, a version of the JS printer should be created that uses "Write Item" IR. That might be faster than both the Rust and previous JS versions, but I'm not sure—we shall see. I'm very skeptical that the Rust version will be faster than the JS.

**Comment by dsherret** [comment]
This is basically done now. I just need to get all the tests to run on travis using the rust printer and js printer.

Unfortunately I haven't noticed any performance improvement. If anything, it's slightly slower. That's ok though because this was mostly done to ensure the rust printer is working properly and I found a lot of bugs in it while doing this. Additionally, I improved the JS implementation.
